src/app/pages/data-protection/cloud-backup/cloud-backup-form/cloud-backup-form.component.ts

```diff
@@ -1,18 +1,290 @@
+
+    folder: ['', [Validators.required]],
+    bucket: ['', [Validators.required]],
+    bucket_input: [''],
+  });
+
+  isLoading = false;
   editingTask: CloudBackup;
 
+  bucketOptions$ = of<SelectOption[]>([]);
+
+  fileNodeProvider: TreeNodeProvider;
+  bucketNodeProvider: TreeNodeProvider;
+
+  readonly newOption = newOption;
+  readonly requiredRoles = [Role.CloudBackupWrite];
+
   constructor(
+    private translate: TranslateService,
+    private fb: FormBuilder,
+    private ws: WebSocketService,
+    private cdr: ChangeDetectorRef,
+    private errorHandler: FormErrorHandlerService,
+    private snackbar: SnackbarService,
+    private filesystemService: FilesystemService,
+    private cloudCredentialService: CloudCredentialService,
     private chainedRef: ChainedRef<CloudBackup>,
   ) {
     this.editingTask = chainedRef.getData();
   }
+
+  ngOnInit(): void {
+    this.setFileNodeProvider();
+    this.setBucketNodeProvider();
+
+    this.form.controls.credentials.valueChanges.pipe(untilDestroyed(this)).subscribe((credentials) => {
+      if (credentials) {
+        this.form.controls.folder.enable();
+        this.loadBucketOptions(credentials);
+      } else {
+        this.form.controls.folder.disable();
+        this.form.controls.bucket.disable();
+        this.form.controls.bucket_input.disable();
+      }
+    });
+
+    this.form.controls.bucket.valueChanges.pipe(untilDestroyed(this)).subscribe((value) => {
+      if (value === newOption) {
+        this.form.controls.bucket_input.enable();
+      } else {
+        this.form.controls.bucket_input.disable();
+      }
+      this.setBucketNodeProvider();
+    });
+
+    this.form.controls.bucket_input.valueChanges.pipe(untilDestroyed(this)).subscribe(() => {
+      this.setBucketNodeProvider();
+    });
+
+    if (this.editingTask) {
+      this.setTaskForEdit();
+    }
+  }
+
+  loadBucketOptions(credentialId: number): void {
+    this.isLoading = true;
+    this.cloudCredentialService.getBuckets(credentialId)
+      .pipe(untilDestroyed(this))
+      .subscribe({
+        next: (buckets) => {
+          const bucketOptions = buckets.map((bucket) => ({
+            label: bucket.Name,
+            value: bucket.Path,
+            disabled: !bucket.Enabled,
+          }));
+          bucketOptions.unshift({
+            label: this.translate.instant('Add new'),
+            value: newOption,
+            disabled: false,
+          });
+          this.bucketOptions$ = of(bucketOptions);
+          this.isLoading = false;
+          this.form.controls.bucket.enable();
+          this.form.controls.bucket_input.disable();
+          this.cdr.markForCheck();
+        },
+        error: () => {
+          this.isLoading = false;
+          this.form.controls.bucket.disable();
+          this.form.controls.bucket_input.enable();
+          this.cdr.markForCheck();
+        },
+      });
+  }
+
+  getBucketsNodeProvider(): TreeNodeProvider {
```

Comment:
Would be good to fix this code duplication, you can add comment for that 

<img width="1191" alt="Screenshot 2024-04-12 at 16 30 46" src="https://github.com/truenas/webui/assets/22980553/33d23532-7db0-4f2e-988c-982c719f0cdd">